### PR TITLE
feature(cody): Data types for encoding LLM Model config via Sourcegraph site config

### PR DESCRIPTION
**Pull request by chrsmith**
This PR is part of the epic space opera that is [Server-side Cody Model Selection](https://linear.app/sourcegraph/project/server-side-cody-model-selection-cca47c48da6d) and friends, such as [Cody with Self-hosted Models](https://linear.app/sourcegraph/project/cody-with-self-hosted-models-39edcf552a64). It doesn't add any user-facing functionality, but rather introduces the foundational data types and conversion logic so we can integrate the new "LLM Model Configuration" code into the Sourcegraph instance.

## Overview

The `internal/modelconfig` package defines the data types and validation logic for LLM Model Configuration. The goal is to provide an ultra-flexible way for a Sourcegraph backend to represent LLM configuration data, so that it can rely on it for registering its own completions API endpoints as well as send that data to Cody clients so that they can understand what LLM models the Sourcegraph backend supports.

To date, all of this LLM configuration data has been bundled into the `types.ModelConfiguration` data type. (`internal/modelconfig/types`) It contains all of the "Providers", "Models", etc. that the Sourcegraph instance understands.

However, this data type is not easy to manipulate. It is designed to make it easy to look up data when we need to process an LLM completion request, or to easily consumable by Cody clients. But there isn't a good way for a Sourcegraph admin to "modify" or "patch" that object.

This PR introduces a new data type that is conceptually the same -- a bunch of LLM model configuration -- but is meant to be expressed by a Sourcegraph administrator manually configuring how Cody should work on their instance.

### Introducing `SiteModelConfiguration`.

Inside `internal/modelconfig/types/documents.go` we define a new type, `SiteModelConfiguration`. This is intended to be encoded directly into the site configuration (`site.schema.json`) and be reasonably easy for a non-expert to configure.

@slimsag and I spent a lot of time talking through the various scenarios that we want to enable, and various tradeoffs for how to represent the data. This is what we wound up with and is able to serve all known requirements. (As far as we know for now. I hope. 😅)

While it isn't acceptable as a valid form of documentation (which I'll produce once the end-to-end of this is all working, and I can show how things come together), I'd suggest you take a look at `internal/modelconfig/validation_test.go` to understand how to think about this new data type. (As well as all the doc comments on the `SiteModelConfiguration` type itself.)

### Converting the existing Site Configuration

In order to provide some more context for how we expect to _use_ these data types, I've included some related functionality in this PR as well.

This PR adds a new `cmd/frontend/internal/modelconfig` package. This is where we will put the Sourcegraph instance-specific logic for reading, updating, and serving the LLM model configuration data.

For now, it just contains utility types and one meaningful function:

```go
// convertCompletionsConfig converts the supplied Completions configuration blob (the Cody Enterprise configuration data)
// into the newer types.SiteModelConfiguration structure.
//
// Assumes that the supplied completions object is valid, and contains all the required settings. e.g. the site admin
// can leave some things blank, but `conf/computed.go`'s `GetCompletionsConfig()` will fill the Endpoint and related
// fields with meaingful defaults.
func convertCompletionsConfig(completionsCfg *conftypes.CompletionsConfig) *types.SiteModelConfiguration {
```

In order to refactor our current completions APIs to use the new `modelconfig` data types, we first need to safely read existing Cody Enterprise configuration data and represent it in this new data structure.

The `convertCompletionsConfig` function loads the `CompletionsConfig` data in the site configuratg (i.e. [all this stuff](https://sourcegraph.com/docs/cody/clients/enable-cody-enterprise) and converts it into a `SiteModelConfiguration`.

This part is _really important_ to get right, since a mistake here would just be a regression for any Cody Enterprise customer that upgrades their Sourcegraph instance. (After I continue refactoring the completions API codepath.)

We'll be able to do more meaningful testing as we land the next few PRs, if you look at `cmd/frontend/internal/modelconfig/siteconfig_completions_test.go` you can get a sense for how the existing Cody Enterprise configuration data gets surfaced in the new data types.

## Next Steps

After this PR lands, the next step is to continue building out the `cmd/frontend/internal/modelconfig` package. Specifically, to expose the `types.ModelConfiguration` data so that the Completions APIs (`cmd/frontend/internal/completions`) can actually _use_ this data for making LLM requests, either via BYOK or to Cody Gateway.

This will involve some super-annoying to write code that "merges" the `types.ModelConfiguration` data that we'll fetch from Cody Gateway and/or is embedded in the Sourcegraph instance with the Sourcegraph admin supplied `types. types.SiteModelConfiguration`.

And... FINALLY!!!! at long last, once _that_ is done, it's just a matter of updating the Completions handlers to use these new data types, and enjoy the strongly typed and easily extendable `ServerSideProviderConfig` configuration objects.

## Test plan

Lots of new unit tests, but none of this is surfaced to end users. YET.

### CODE REVIEW

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+		return nil
+	}
+
+	// convertToModelRef convert the format of model defined in the configuration to
+	// a ModelRef, e.g.
+	// "foo" => "${completionsCfg.provider}::unknown::foo"
+	// "foo/bar" => "foo::unknown::bar"
+	//
+	// And aws-bedrock provisioned throughput... is handled in its own way.
+	convertToModelRef := func(cfgModelID string) string {
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:28]: I think that we also need a special case in this function for Azure.

For context:

* When admins upgrade, if they have `"provider": "azure-openai"` then they will be warned at the top of the page `when using azure-openai provider its mandatory to set both completions.azureChatModel and completions.azureCompletionModel for proper LLM Token usage` - this was implemented in https://github.com/sourcegraph/sourcegraph/pull/63100/files#diff-c4de0420e936b4e62127fd642bf96984f80610be72f612012d66fc3014b3a502R78 - and they will not have `completionsCfg.azureChatModel` or `completionsCfg.azureCompletionModel` set yet.
* `cfgModelID` here would refer to the Azure deployment name -- not the actual model.

The way I think you should handle this in this PR:

1. When `convertToModelRef` is being called for the `azure-openai` provider, use the values of `completionsCfg.AzureChatModel` instead of `completionsCfg.ChatModel`, same for the ocmpletions model.
2. For fastchat .... it seems we didn't consider this case. Introduce a new AzureFastChatModel site config field and respect that?

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			Chat:           types.ModelRef(chatModelRef),
+			CodeCompletion: types.ModelRef(autocompleteModelRef),
+			FastChat:       types.ModelRef(fastModelRef),
+		},
+	}
+
+	// Account for the way we encoded the AWS Provisioned Throughput ARN into the site config.
+	// We overloaded the model name with both the actual model name and an ARN. So if we see
+	// that, add the server-side configuration data.
+	if completionsCfg.Provider == "aws-bedrock" {
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:123]: I wonder if we need special handling like this to grab the Azure deployment name out, too?

File: internal/modelconfig/types/documents.go [hunk 3]
```diff
@@ -19,3 +19,93 @@ type ModelConfiguration struct {
...
+	//                  "accessToken": "secret",
+	//                  "endpoint": "https://llm-models-r-us.com/"
+	//              }
+	//          }
+	//      },
+	//      {
+	//          "id": "anthropic",
+	//          "serverSideConfig": {
+	// 	            "awsBedrockConfig": {
+	// 		            "region": "us-west-2",
```
Comment (slimsag) [inline internal/modelconfig/types/documents.go:55]:
```suggestion
	// 		    "region": "us-west-2",
```

File: internal/modelconfig/types/documents.go [hunk 4]
```diff
@@ -19,3 +19,93 @@ type ModelConfiguration struct {
...
+	// expanded context window from what Sourcegraph supplied:
+	//
+	// ```
+	// "modelOverrides": [
+	//     ...
+	//     {
+	//         "modelRef": "anthropic::2023-06-01::claude-3-sonnet",
+	//         "contextWindow": {
+	//		       "maxInputTokens": 200000,
+	//	           "maxOutputTokens": 20000
```
Comment (slimsag) [inline internal/modelconfig/types/documents.go:95]:
```suggestion
	//	       "maxOutputTokens": 20000
```

Comment (slimsag) [reply internal/modelconfig/types/documents.go:95]: you've got a bunch of mixed spaces and tabs in these docstrings, maybe easier for you to fix them in-editor

File: internal/modelconfig/types/documents.go [hunk 5]
```diff
@@ -19,3 +19,93 @@ type ModelConfiguration struct {
...
+	//         }
+	//     }
+	//     ...
+	// ]
+	// ```
+	//
+	// If the ModelRef is unknown (e.g. the Sourcegraph-supplied model was removed
+	// via the SourcegraphModelConfig's deny list.) In that case, all model settings
+	// need to be supplied, either explicitly or by the provider's DefaultModelConfig.
+	ModelOverrides []ModelOverride `json:"modelOverrides"`
```
Comment (slimsag) [inline internal/modelconfig/types/documents.go:105]: Just FYI - this `ModelOverride` struct is currently missing the ability to use 'recommended' settings as we talked about for a specific model key, e.g.:

```
        // Replaces any models, and may just add new ones if the mref is unique.
        "modelOverrides": [
            {
                "mistral::v1::mixtral-8x22b-instruct": {
                    "useRecommendSettings": true,
                },
                ...
            }
        ]
```

I can work on this in a follow-up PR, but will be something I want for self-hosted models

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go (hunk 1 above)
Comment (chrsmith) [reply cmd/frontend/internal/modelconfig/siteconfig_completions.go:28]: > I think that we also need a special case in this function for Azure.

Yes, absolutely. In the latest set of changes stacked on top of this, I've made that change.

However, I do need to update it to honor `azureChatModel`/`azureCompletionModel`... what is the difference between that and `chatModel`/`completionModel`? For the Azure case the `azureCompletionModel` is the _actual_ LLM model name, whereas the `compoetionModel` is the "Azure deployment name"?

File: internal/modelconfig/types/documents.go (hunk 5 above)
Comment (chrsmith) [reply internal/modelconfig/types/documents.go:105]: Ah yes. I ended up just having reasonable-ish defaults for everything but `CompletionWindow`. Will call that out in the next update.

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go (hunk 1 above)
Comment (slimsag) [reply cmd/frontend/internal/modelconfig/siteconfig_completions.go:28]: Yes, with Azure, `chatModel`, `fastChatModel`, and `completionModel` are all the "Azure deployment name" - a unique identifier that comes from the customer's azure deployment - a grave mistake we made when introducing that provider.

To try to correct it, we introduced `azureChatModel`, `azureCompletionModel` which are the actual model names like `gpt-35-turbo-instruct` and not unique to the customer's deployment.

File: internal/modelconfig/types/documents.go (hunk 5 above)
Comment (slimsag) [reply internal/modelconfig/types/documents.go:105]: Just to be clear, the goal here is for me Stephen to be able to express this:

If a site admin is configuring `mistral::v1::mixtral-8x22b-instruct` specifically (no other mrefs, just that one in specific - hard-coded in our Go code somewhere), and is a self-hosted model customer, let the site admin say "use whatever configuration Stephen suggests for all fields here with this mref".

**Review by slimsag — APPROVED** [review]
Excellent work

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+func (k defModelKind) String() string {
+	switch k {
+	case defModelChatModel:
+		return "chat"
+	case defModelCompletion:
+		return "completion"
+	case defModelFastChatModel:
+		return "fast-chat"
+	default:
+		return "unknown"
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:34]: do we actually need to express an unknown case in some real scenario? Otherwise I might suggest panic()

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+		}
+
+		// Create ProviderOverride if we haven't seen this provider before.
+		// We need to remap the provider ID if it is referring to an API Provider and not
+		// a Model Provider.
+		var effectiveProviderID string
+		switch legacyModelRef.provider {
+		case "aws-bedrock":
+			effectiveProviderID = "anthropic"
+		case "azure-ai":
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:171]:
```suggestion
		case "azure-openai":
```

Unless I'm mistaken this is referring to the legacy model provider from the site config

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+		}
+	}
+
+	defaultModels := types.DefaultModels{
+		Chat:           configuredModels[defModelChatModel].ModelRef,
+		CodeCompletion: configuredModels[defModelCompletion].ModelRef,
+		FastChat:       configuredModels[defModelFastChatModel].ModelRef,
+	}
+
+	// Now lineraize those maps.
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:310]:
```suggestion
	// Now linearize those maps.
```

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+	return &serverSideConfig
+}
+
+// convertCompletionsConfig converts the supplied Completions configuration blob (the Cody Enterprise configuration data)
+// into the newer types.SiteModelConfiguration structure.
+//
+// Assumes that the supplied completions object is valid, and contains all the required settings. e.g. the site admin
+// MAY leave some things blank, but `conf/computed.go`'s `GetCompletionsConfig()` will provide defaults for all the
+// values. So even if the site admin didn't provide the `Endpoint` field, we expect it to be present.
+func convertCompletionsConfig(completionsCfg *conftypes.CompletionsConfig) (*types.SiteModelConfiguration, error) {
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions.go:134]: minor comment: this function surface-level looks super complex, I feel like some portions of it could be broken out into more readable function names.

**Review by slimsag — APPROVED** [review]
Great

File: internal/modelconfig/types/documents.go (hunk 4 above)
Comment (chrsmith) [reply internal/modelconfig/types/documents.go]: Sorry about that, fixed.

File: internal/modelconfig/types/documents.go (hunk 5 above)
Comment (chrsmith) [reply internal/modelconfig/types/documents.go:105]: Gotcha. Then yes, that is something that isn't accounted for in these Go types. (And perhaps that would be something that "only exists" in the Site Config projection? And those "useRecommendedSettings" get applied when reading the site config and converting it into the `modelconfig` types?)

For reference, here's how this works in this particular PR:

- We define a `DefaultModelConfig` for each `ProviderOverride`.
- We do not provide a default for `ContextWindow` (which is definitely required), because we honor `CompletionModelMaxTokens` and related.

```go
	// Generic defaults.
	defaultModelConfig := types.DefaultModelConfig{
		Capabilities: []types.ModelCapability{
			types.ModelCapabilityAutocomplete,
			types.ModelCapabilityChat,
		},
		Category: types.ModelCategoryBalanced,
		Status:   types.ModelStatusStable,
		Tier:     types.ModelTierEnterprise,

		// IMPORTANT: The default model config contains an invalid
		// context window (0, 0). The ModelOverrides MUST set the
		// expected values.
	}
```

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go (hunk 9 above)
Comment (chrsmith) [reply cmd/frontend/internal/modelconfig/siteconfig_completions.go:179]: > this function surface-level looks super complex

OMFG yes.

As I later added more tests, and accounted for more corner cases, I ended up rewriting most of it. While the function is _still_ super long and surprisingly complicated, it is hopefully easier to follow.

**Comment by chrsmith** [comment]
@slimsag would you mind taking another look before I submit?

- Merged in the latest changes from `main`
- Updated the code to account for `AzureChatModel` and friends, added a unit test
- Added a `modelconfig.SanitizeResourceName(string) string` function. Since it'll come in handy.

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go (hunk 8 above)
Comment (chrsmith) [reply cmd/frontend/internal/modelconfig/siteconfig_completions.go:355]: Fixed.

File: cmd/frontend/internal/modelconfig/siteconfig_completions.go (hunk 7 above)
Comment (chrsmith) [reply cmd/frontend/internal/modelconfig/siteconfig_completions.go:216]: Yes. Good catch.

This stanza is actually kinda redundant, since we do this same "provider remapping" inside of `parseLegacyModelRef(...)`. But given how subtle it is, I figured it was worth keeping in even if it is a no-op/duplicate. (Of course, it doesn't help if it uses the wrong enum value... 😭)

```go
	// AzureOpenAI
	case conftypes.CompletionsProviderNameAzureOpenAI:
		// AzureOpenAI (at least for v5.3) only supports openai-provided models.
		// However, the site configuration requires you to specify the "Azure Deployment ID"
		// in place of the model name. And to later diambiguate this, admins could optionally
		// supply more meaningful model names. (Although the "deployment ID" is what is still
		// required to be used when making the API call.)
		providerID = "openai"
```

File: cmd/frontend/internal/modelconfig/siteconfig_completions_test.go [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+			assert.True(t, azureProviderConfig.UseDeprecatedCompletionsAPI)
+
+			// ModelOverrides
+			// For AzureOpenAI, we verify that the "ModelName" and "ModelID" are set independently, based on
+			// the configuration data.
+			require.Equal(t, 3, len(siteModelConfig.ModelOverrides))
+			{
+				// BUG: The "model name alias" configuration doesn't apply to FastChat, hence the awkward ModelID.
+				m := siteModelConfig.ModelOverrides[0]
+				assert.EqualValues(t, "openai::unknown::azure-deployment-id_-__54e1331b-4ccc-44b8-b9ef-59c6b050fdb8_", m.ModelRef)
```
Comment (slimsag) [inline cmd/frontend/internal/modelconfig/siteconfig_completions_test.go:280]: Leaving this comment as a reminder for myself to add an `azureFastChatModel` and fix this. I'll send a PR for this

Comment (chrsmith) [reply cmd/frontend/internal/modelconfig/siteconfig_completions_test.go:280]: > as a reminder for myself to add an azureFastChatModel and fix this.
🙏 I'd offer to just fix this myself, if I didn't already have like several thousand lines of code that needs to be teased out into PRs already 😬 🙇 .

> to not say (optional) anymore. we should just require these fields.
💯 🙏

**Comment by slimsag** [comment]
@chrsmith the only thing I see missing here - and maybe we can add it as a follow-on later, is the deployment names like `"azure-deployment-id - {7d1f5676-189d-4a97-9dd9-10ae187aba99}"` are going to be needed by the actual azure backend Go client. I don't see anywhere here that we store them in the Go data structures, but maybe I missed it?

**Review by slimsag — APPROVED** [review]
lgtm :shipit:

**Comment by chrsmith** [comment]
> @chrsmith the only thing I see missing here - and maybe we can add it as a follow-on later, is the deployment names like "azure-deployment-id - {7d1f5676-189d-4a97-9dd9-10ae187aba99}" are going to be needed by the actual azure backend Go client. I don't see anywhere here that we store them in the Go data structures, but maybe I missed it?

@slimsag great question.

This will make more sense when I send out the PRs that refactor how the actual completion providers work, but I believe we have all the information we need on the `types.Model`.

The `Model.ID` is the opaque ID just used to uniquely identify the model, e.g. "gpt-4o". The _meaningful_ value that we pass to the API Provider for that particular model however is in the `Model.ModelName` field. So as long _that_ value has the "azure-deployment-id - {7d1f5676-189d-4a97-9dd9-10ae187aba99}" string, then things should work just fine.

And this is the behavior we have today. It's just that, we treat `Model.ID == Model.ModelName`, instead of having these as two distinct concepts. (Among other changes introduced with the new types.)

**Comment by slimsag** [comment]
Sounds good, as long as the Go client has a way to get that exact `"azure-deployment-id - {7d1f5676-189d-4a97-9dd9-10ae187aba99}"` string as it was originally written, should be fine.
